File: scripts/official-media.py
import argparse
import hashlib
import io
import json
import time
import urllib.error
import urllib.request
from pathlib import Path
from PIL import Image

root = Path(__file__).resolve().parent.parent
parser = argparse.ArgumentParser()
parser.add_argument('--offline', action='store_true')
args = parser.parse_args()
cache = root / '.cache' / 'official'
output = root / 'dist' / 'media' / 'official'
cache.mkdir(parents=True, exist_ok=True)
output.mkdir(parents=True, exist_ok=True)
Image.MAX_IMAGE_PIXELS = 16_000_000
items = json.loads((root / 'data' / 'official-media.json').read_text())

for item in items:
    cached = cache / (item['id'] + '.source')
    if cached.exists():
        raw = cached.read_bytes()
    elif args.offline:
        raise RuntimeError('Missing offline official image: ' + item['id'])
    else:
        raw = None
        for attempt in range(3):
            try:
                request = urllib.request.Request(
                    item['url'],
                    headers={
                        'User-Agent': 'Mozilla/5.0 (compatible; CQUPT-Running-Archive/1.0; +https://github.com/yhan-sun/cqupt-zyfx)',
                        'Referer': 'https://mp.weixin.qq.com/'
                    }
                )
                with urllib.request.urlopen(request, timeout=30) as response:
                    candidate = response.read(5_000_001)
                if len(candidate) < 1000 or len(candidate) > 5_000_000:
                    raise ValueError('Unexpected image size: ' + item['id'])
                if hashlib.sha256(candidate).hexdigest() != item['sha256']:
                    raise ValueError('Official image source changed; review required: ' + item['id'])
                raw = candidate
                cached.write_bytes(raw)
                break
            except (urllib.error.URLError, TimeoutError, ConnectionError) as error:
                if attempt == 2:
                    raise
                delay = 3 * (attempt + 1)
                print('Retry official image', item['id'], 'in', delay, 'seconds:', error, flush=True)
                time.sleep(delay)
        if raw is None:
            raise RuntimeError('Could not retrieve official image: ' + item['id'])
    if hashlib.sha256(raw).hexdigest() != item['sha256']:
        raise ValueError('Invalid cached official image: ' + item['id'])
    with Image.open(io.BytesIO(raw)) as probe:
        probe.verify()
    with Image.open(io.BytesIO(raw)) as image:
        image.load()
        if image.size != (item['width'], item['height']):
            raise ValueError('Unexpected dimensions: ' + item['id'])
        rgb = image.convert('RGB')
        rgb.thumbnail((1200, 1200), Image.Resampling.LANCZOS)
        rgb.save(output / (item['id'] + '.webp'), 'WEBP', quality=82, method=6)
        item['builtWidth'], item['builtHeight'] = rgb.size
    print('Official image', item['id'], item['width'], 'x', item['height'], 'SHA-256 verified', flush=True)

(output / 'credits.json').write_text(json.dumps(items, ensure_ascii=False, indent=2))

File: scripts/official-media.mjs
import { fileURLToPath } from 'node:url';
import { access, readFile, writeFile } from 'node:fs/promises';
import path from 'node:path';
import sharp from 'sharp';
import {
  downloadBytes,
  ensureDirectory,
  readJson,
  retry,
  root,
  sha256,
  writeJson
} from './media-utils.mjs';

const MAX_BYTES = 5_000_000;
const MAX_PIXELS = 16_000_000;

function imageInput(raw) {
  return sharp(raw, { limitInputPixels: MAX_PIXELS });
}

async function validateImage(raw, item) {
  const image = imageInput(raw);
  const metadata = await image.metadata();
  if (metadata.width !== item.width || metadata.height !== item.height) {
    throw new Error(`Unexpected dimensions: ${item.id}`);
  }
  await image.removeAlpha().raw().toBuffer();
}

async function fileExists(filename) {
  try {
    await access(filename);
    return true;
  } catch (error) {
    if (error.code === 'ENOENT') return false;
    throw error;
  }
}

async function getSource(item, offline) {
  const cache = path.join(root, '.cache', 'official');
  const cached = path.join(cache, `${item.id}.source`);
  await ensureDirectory(cache);

  if (await fileExists(cached)) return readFile(cached);
  if (offline) throw new Error(`Missing offline official image: ${item.id}`);

  return retry(
    async () => {
      const raw = await downloadBytes(item.url, {
        headers: { 'User-Agent': 'Mozilla/5.0', Referer: 'https://mp.weixin.qq.com/' },
        maxBytes: MAX_BYTES,
        timeoutMs: 30_000
      });
      if (raw.length < 1000) throw new Error(`Unexpected image size: ${item.id}`);
      if (sha256(raw) !== item.sha256) throw new Error(`Official image source changed; review required: ${item.id}`);
      await validateImage(raw, item);
      await writeFile(cached, raw);
      return raw;
    },
    {
      attempts: 3,
      shouldRetry: error => error.retryable === true
    }
  );
}

async function renderImage(item, raw, output) {
  const result = await imageInput(raw)
    .removeAlpha()
    .resize({ width: 1200, height: 1200, fit: 'inside', withoutEnlargement: true, kernel: sharp.kernel.lanczos3 })
    .webp({ quality: 82, effort: 6 })
    .toBuffer({ resolveWithObject: true });
  await writeFile(path.join(output, `${item.id}.webp`), result.data);
  return { ...item, builtWidth: result.info.width, builtHeight: result.info.height };
}

export async function buildOfficialMedia({ offline = false } = {}) {
  const items = await readJson('data/official-media.json');
  const output = path.join(root, 'dist', 'media', 'official');
  await ensureDirectory(output);
  const processed = [];

  for (const item of items) {
    const raw = await getSource(item, offline);
    if (sha256(raw) !== item.sha256) throw new Error(`Invalid cached official image: ${item.id}`);
    await validateImage(raw, item);
    processed.push(await renderImage(item, raw, output));
    console.log(`Official image ${item.id} ${item.width} x ${item.height} SHA-256 verified`);
  }
  await writeJson('dist/media/official/credits.json', processed);
}

if (process.argv[1] && path.resolve(process.argv[1]) === fileURLToPath(import.meta.url)) {
  await buildOfficialMedia({ offline: process.argv.includes('--offline') });
}
